### PR TITLE
Convert all throws into Results

### PR DESCRIPTION
**Pull request by lucperkins**
A small quality of life thing but one that should have some ergonomic benefits across our Actions. Basically makes this lib more like Rust. No more try/catch, just `Result<T, string>` (where the string is error text). There _are_ still try/catch blocks in this library, of course, but nothing here _throws_ errors. I've also streamlined a few other things, like assigning main and post hooks to `IdsToolbox`.

You can see an example in DeterminateSystems/nix-installer-action#81

### CODE REVIEW

File: src/linux-release-info.ts [hunk 1]
```diff
@@ -153,68 +155,64 @@ function getOsInfo(): OsInfo {
...
   let fileData = null;
 
   for (const osReleaseFile of fileList) {
     try {
-      if (options.debug) {
-        /* eslint-disable no-console */
-        console.log(`Trying to read '${osReleaseFile}'...`);
-      }
+      if (options.debug)
+        actionsCore.debug(`Trying to read '${osReleaseFile}'...`);
```
Comment (grahamc) [inline src/linux-release-info.ts:164]:
```suggestion
      if (options.debug)
        actionsCore.debug(`Trying to read '${osReleaseFile}'...`);
```

-1 on the braceless if's

**Review by grahamc — COMMENTED** [review]
I've gotta say I'm not real thrilled about this PR, because it seems weird. Is this a thing that a lot of JS projects are doing? I don't want to be weird here.

**Comment by lucperkins** [comment]
Not a lot of projects do it, no, but throws and try/catch blocks are completely and utterly inscrutable and untraceable in JS. Anything can throw an error at any time and there's no guarantee that that error is properly handled or even surfaced in a way that indicates where it came from. This provides us with full control of the error handling "stack" in a way that JS on its own does not afford.

**Comment by lucperkins** [comment]
And not a lot of projects do it because this is widely seen as one of those this-is-fine.png aspects of working with JS. Just a kind of brute fact we all have to suffer under. It's up to you but I'd like to see us reject that "fact."

**Comment by lucperkins** [comment]
The library is much more well structured now, so I'm less concerned about stray `throws`.

**Pull request closed without merging**